### PR TITLE
feat(sql): add greatest and least functions

### PR DESCRIPTION
**Pull request by nwoolmer**
Adds `greatest(V)` which will give the highest number out of its arguments.
Adds `least(V)`, which gives the lowest number of its arguments

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 108 / 126 (85.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/LeastNumericFunctionFactory.java|54|63|85.71%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/GreatestNumericFunctionFactory.java|54|63|85.71%|

File: core/src/main/java/io/questdb/griffin/engine/functions/math/LeastNumericFunctionFactory.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+    @Override
+    public Function newInstance(
+            int position,
+            ObjList<Function> args,
+            IntList argPositions,
+            CairoConfiguration configuration,
+            SqlExecutionContext sqlExecutionContext
+    ) throws SqlException {
+        final int[] counters = new int[11];
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/LeastNumericFunctionFactory.java:59]: 11 is a bit harsh and also, lets make this thread-local!

File: core/src/main/java/io/questdb/griffin/engine/functions/math/GreatestNumericFunctionFactory.java [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+        if (counters[ColumnType.SHORT] > 0) {
+            return new CastLongToShortFunctionFactory()
+                    .newInstance(position,
+                            new ObjList<>(new GreatestLongRecordFunction(new ObjList<>(args), argPositions)),
+                            null, configuration, sqlExecutionContext);
+        }
+
+        if (counters[ColumnType.BYTE] > 0) {
+            return new CastLongToByteFunctionFactory()
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/math/GreatestNumericFunctionFactory.java:121]: may be we can be more frugal here and bypass cast function factory call by skipping directly to cast function?

CastFactory always uses args.get(0), we don't need to create an extra obj list

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 108 / 126 (85.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/LeastNumericFunctionFactory.java|54|63|85.71%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/GreatestNumericFunctionFactory.java|54|63|85.71%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 126 / 142 (88.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/LeastNumericFunctionFactory.java|61|69|88.41%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/GreatestNumericFunctionFactory.java|61|69|88.41%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToByteFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToShortFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToFloatFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToIntFunctionFactory.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 126 / 142 (88.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/LeastNumericFunctionFactory.java|61|69|88.41%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/math/GreatestNumericFunctionFactory.java|61|69|88.41%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToByteFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToShortFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastDoubleToFloatFunctionFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/cast/CastLongToIntFunctionFactory.java|1|1|100.00%|

**Pull request merged**